### PR TITLE
chore: Remove apply_generic, use unary_elementwise

### PR DESCRIPTION
**Pull request by MarcoGorelli**
This is what I was talking about yesterday

It seems to me that `apply_generic` and `unary_elementwise` overlap? There's an extra little fastpath in the first one for when there's no nulls (which I ported over), but other than that, is it necessary to have both?

### CODE REVIEW

File: crates/polars-core/src/chunked_array/mod.rs [hunk 1]
```diff
@@ -75,8 +75,8 @@ pub type ChunkLenIter<'a> = std::iter::Map<std::slice::Iter<'a, ArrayRef>, fn(&A
 ///
 /// ```rust
 /// # use polars_core::prelude::*;
-/// fn apply_cosine_and_cast(ca: &Float32Chunked) -> Float64Chunked {
-///     ca.apply_values_generic(|v| v.cos() as f64)
+/// fn apply_cosine_and_cast(ca: &Float32Chunked) -> Float32Chunked {
+///     ca.apply(|v| v.cos())
```
Comment (MarcoGorelli) [inline crates/polars-core/src/chunked_array/mod.rs]: the description above says

> This allows us to apply closures to the data and collect the results to a [`ChunkedArray`] of the same type `T`.

so I think a plain `apply` might be a better example here?

File: crates/polars-core/src/chunked_array/mod.rs [hunk 2]
```diff
@@ -75,8 +75,8 @@ pub type ChunkLenIter<'a> = std::iter::Map<std::slice::Iter<'a, ArrayRef>, fn(&A
 ///
 /// ```rust
 /// # use polars_core::prelude::*;
-/// fn apply_cosine_and_cast(ca: &Float32Chunked) -> Float64Chunked {
-///     ca.apply_values_generic(|v| v.cos() as f64)
+/// fn apply_cosine_and_cast(ca: &Float32Chunked) -> Float32Chunked {
+///     ca.apply_values(|v| v.cos())
```
Comment (MarcoGorelli) [inline crates/polars-core/src/chunked_array/mod.rs:79]: the text above says

> This allows us to apply closures to the data and collect the results to a [`ChunkedArray`] of the same type `T`.

so I think `apply_values` might be a better example to show?

File: crates/polars-core/src/chunked_array/ops/arity.rs [hunk 3]
```diff
@@ -74,10 +74,17 @@ where
...
 {
-    let iter = ca
-        .downcast_iter()
-        .map(|arr| arr.iter().map(&mut op).collect_arr());
-    ChunkedArray::from_chunk_iter(ca.name(), iter)
+    if ca.null_count == 0 {
+        let iter = ca
+            .downcast_iter()
+            .map(|arr| arr.values_iter().map(|x| op(Some(x))).collect_arr());
+        ChunkedArray::from_chunk_iter(ca.name(), iter)
```
Comment (MarcoGorelli) [inline crates/polars-core/src/chunked_array/ops/arity.rs:81]: porting this over from `apply_generic`

File: crates/polars-core/src/chunked_array/ops/arity.rs [hunk 4]
```diff
@@ -74,10 +74,17 @@ where
     F: UnaryFnMut<Option<T::Physical<'a>>>,
     V::Array: ArrayFromIter<<F as UnaryFnMut<Option<T::Physical<'a>>>>::Ret>,
 {
-    let iter = ca
-        .downcast_iter()
-        .map(|arr| arr.iter().map(&mut op).collect_arr());
-    ChunkedArray::from_chunk_iter(ca.name(), iter)
+    if ca.null_count == 0 {
```
Comment (ritchie46) [inline crates/polars-core/src/chunked_array/ops/arity.rs:77]: can this use `has_nulls`. We want to optimize this check later.

**Pull request merged**